### DISCUSSION TITLE
warm_start  option in native  julia EquationSearch

### DISCUSSION THREAD

**Discussion by AxelJessner** [MilesCranmer/PySR#328] [Q&A] [answered]
I am interested in fitting spectra of radio pulsars with the simplest possible function and started with using the PySR version. I find it however a lot  more convenient and efficient  to use the native  julia version, picking up a few julia skills on the way.  

Pulsar spectra span a dynamic range of several orders of magnitude and therefore require a chi^2 like loss function which I have specified following the examples given in the docs.  
lossxyw(x,y,w) =((x-y)*w)^2;
 with w=1/sigma

 Here is a typical spectrum with the fit by EquationSearch: 
![B2021](https://github.com/MilesCranmer/PySR/assets/133158048/c261d839-026a-4962-a45c-33c66d8bfecd)
The fit given as fit1(x1)=(16717.066101066815 / (1349.00938061052 + x1)) ^ 1.9131948224707083
The fit is good and the algorithm converges quite reliably on the _form_ of the function, but  understandably enough with variable results for the constants in different runs, and also with variable goodness of fit. 
I am using the following options for my call of EquationSearch:
```julia
options = Options(;
    binary_operators=[+, -, *, /,^], 
    unary_operators=[log,exp], 
    complexity_of_operators=[(/) => 2] , 
    complexity_of_variables=2,
    constraints=[(^)=>(-1,1), (*) => (5,5)],
    nested_constraints=[ (/) => [(/) => 0], exp => [exp => 0],log => [log => 0]], 
    elementwise_loss=lossxyw,  
    enable_autodiff=true,
    maxsize=20,
    save_to_file=false,
    optimizer_iterations=16,
    ncycles_per_iteration=2000
)
hof = EquationSearch(X, y; weights=w, options=options,  niterations=100)
```
I would now start a subsequent iteration with different option where the algorithm should focus on refining the constants
in order to find the functions with the smallest loss.

my idea has been to use 
```julia
options2 = Options(;
    binary_operators=[+, -, *, /,^], 
    unary_operators=[log,exp], 
    complexity_of_operators=[(/) => 2] , 
    complexity_of_variables=2,
    constraints=[(^)=>(-1,1), (*) => (5,5)],
    nested_constraints=[ (/) => [(/) => 0], exp => [exp => 0],log => [log => 0]], 
     elementwise_loss=lossxyw,  
    enable_autodiff=true,
    maxsize=20,
    save_to_file=false,
    optimizer_iterations=32,
    optimize_probability=0.2,
    warm_start=true,
    ncycles_per_iteration=500
)
```
for niterations=16.

Unfortunately the warm_start=true option from PySR  is not supported and I do not understand how to feed the previous state into a subsequent run. Is there any simple way to achieve that for the native julia implementation?

Another, but unrelated problem is the apparent inability to recover Lorentzian functions like y(x)=1 / (1 +1.0e-4* x^2)    even when data are generated directly from them for tests. SR comes up with arbitrary complex solutions, but doesn't hit on the simple ones. It doesn't matter if log and exp are allowed here or not.  Is there a way to tweak the options so that it might evaluate them too, but without biasing against other simple alternatives?

A few hints would be greatly appreciated for me as a novice here!

Thanks in advance

  Axel

**Comment by MilesCranmer** [comment]
Hi @AxelJessner,

For question 1: Yes, in the Julia side of things, you can use warm starts by setting `return_state=true` in the `Options` object, and then `EquationSearch` will return an extra state variable.

For example:

```julia
options = Options(; kws..., return_state=true)
hof, state_info = EquationSearch(X, y; options)

...

hof2, state_info2 = EquationSearch(X, y; options, saved_state=(hof, state_info))
```

I understand it's a bit less intuitive than the PySR side of things, so I am actively trying to think up an improved API. If you have any suggestions, the thread is here: https://github.com/MilesCranmer/SymbolicRegression.jl/issues/187.

Excited to see what you find with pulsars!!
Cheers,
Miles

**Reply by MilesCranmer** [reply]
Btw, for question 2. (the Lorentzian), if it's not too much trouble, would you be willing to make a separate thread for it? It helps for discoverability on Google if each question is its own page. No problem if not though! Happy to help either way.

**Reply by AxelJessner** [reply]
Dear Miles,

first of all, let me express my appreciation and gratitude for providing SymbolicRegression.jl and of course for your quick response.

EquationSearch ran as expected after adding the option 'return_state=true' and calling it as advised with e.g.
hof, old_state = EquationSearch(X, y; weights=w, options=options, niterations=100)

It however appears to have modified the structure of hof in some way to make it incompatible with the following call to

best, df = get_best(; hof, options)

that failed with

 best, df = get_best(; hof, options)
ERROR: MethodError: no method matching var"#get_best#1037"(::Vector{Vector{Population{Float64, Float64}}}, ::Options{Int64}, ::typeof(get_best))
Closest candidates are:
  var"#get_best#1037"(::HallOfFame{T, L}, ::Any, ::typeof(get_best)) where {T, L} at D:\E\work\julia\SR2021+51-01.jl:35
Stacktrace:
 [1] top-level scope
   @ REPL[134]:1

where get_best is the function given in the first example in the docs:

function get_best(; hof::HallOfFame{T,L}, options) where {T,L}
    dominating = calculate_pareto_frontier(hof)

    df = DataFrame(;
        tree=[m.tree for m in dominating],
        loss=[m.loss for m in dominating],
        complexity=[compute_complexity(m.tree, options) for m in dominating],
    )

    df[!, :score] = vcat(
        [L(0.0)],
        -1 .* log.(df.loss[2:end] ./ df.loss[1:(end - 1)]) ./
        (df.complexity[2:end] .- df.complexity[1:(end - 1)]),
    )

    min_loss = min(df.loss...)

    best_idx = argmax(df.score .* (df.loss .<= (2 * min_loss)))

    return df.tree[best_idx], df
end

What do I need to do now in order to adapt to the new HallofFame structure?

 I am using SymbolicRegression v0.18.0, DataFrames v1.5.0

I previously used the 'best' result to extracted the function in order to check the goodness of fit and for further processing. This works very well with the following scheme after a successful call to get_best:

eqn = node_to_symbolic(best, options)
 fs="fit1(x1)="*string(eqn)
 eval(Meta.parse(fs))

...

Many thanks in advance for your patience

cheers

  Axel 

P.S.:  Concerning the second question (Lorentzian): I agree and  will create a separate thread when I have done a bit more research into the problem.

**Reply by MilesCranmer** [reply]
Oops, my mistake, sorry about that! It’s actually `(state, hof)` as the output rather than what I wrote above.

Admittedly it needs a better interface and I’m eager to push on https://github.com/MilesCranmer/SymbolicRegression.jl/issues/187 soon

**Reply by AxelJessner** [reply, marked as answer]
Dear  Miles,

thanks a lot, it does work that way indeed!

Please note that the order of arguments needs to be reversed for saved_state in the subsequent calls, too. So for example when the first call is 

 (state1 , hof)  = EquationSearch(X, y; weights=w, options=options,  niterations=10)

 then the second should be 

(state2, hof2) = EquationSearch(X, y; weights=w, options=options2, saved_state=(state1, hof), niterations=16)

Changing options in this way may  actually improve convergence  for some problems. In my case the initial  80-100 iterations were reduced to 10 +16 = 26. 

cheers
   Axel
P.S.: Otherwise I don't see any urgent problem with the interface itself. 
An integration with Measurements.jl and in that way supporting error propagation into the constants would however be of practical use for those who work with  actual measurements. I realise that this is however  quite a tall order and would probably require a major rewrite of lots of code.

**Reply by MilesCranmer** [reply]
Awesome!!

Regarding Measurements.jl integration, maybe it already works? The issue I merged yesterday https://github.com/SymbolicML/DynamicExpressions.jl/pull/33 preserves the same container type. And already, you can provide a different data type than loss type — this is how complex number expressions are optimized.
